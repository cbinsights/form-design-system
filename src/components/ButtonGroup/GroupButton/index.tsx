import React from 'react';
import cc from 'classcat';
import baseElement from 'util/baseElement';
import Icon from 'components/Icon';

const WrapperComponent = ({ children }: { children: JSX.Element }): JSX.Element =>
  children;

export interface GroupButtonProps {
  /** Displays active button style when true */
  isActive?: boolean;

  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link?: JSX.Element;

  /** Text that appears in GroupButton */
  label?: string;

  /** Controls whether button is disabled or not (works for
   * anchor tags as well, visually) */
  disabled?: boolean;

  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon?: Icon;

  /**
   * A component to wrap the entire group button. The
   * `wrapper` must render the children prop passed into it
   * to render the GroupButton.
   */
  Wrapper?: (props: { children: React.ReactNode }) => JSX.Element;

  /** Assigns the first button styling. */
  isFirstButton?: boolean;

  /** Assigns the last button styling. */
  isLastButton?: boolean;

  /** Specifies a URL to link to.  */
  href?: string;

  onClick: () => void;
}

export const GroupButton = (props: GroupButtonProps): JSX.Element => {
  const {
    Wrapper = WrapperComponent,
    isActive,
    Icon,
    label,
    Link,
    isFirstButton,
    isLastButton,
    href,
    ...rest
  } = props;

  const Element = baseElement({ href: href, onClick: true, as: Link });

  const rootClass = cc([
    {
      'groupbtn--disabled': props.disabled,
      'groupbtn--active': isActive,
      'groupbtn--first': isFirstButton,
      'groupbtn--last': isLastButton,
    },
    'groupbtn',
    'border--focus--noTransition',
    'transition--default',
  ]);

  return (
    <Wrapper>
      <Element {...rest} className={rootClass}>
        {Icon && (
          <div className="margin--right--s">
            <Icon size="xs" />
          </div>
        )}
        <span className="groupbtn-label">{label}</span>
      </Element>
    </Wrapper>
  );
};

export default GroupButton;
