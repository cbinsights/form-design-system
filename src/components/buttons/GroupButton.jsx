import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import baseElement from '../../util/baseElement';

export const GroupButton = (props) => {
  const {
    isActive,
    Icon,
    label,
    Link,
    Wrapper,
    isFirstButton,
    isLastButton,
    ...rest
  } = props;

  const Element = baseElement({ href: rest.href, as: Link });

  const rootClass = cx('groupbtn', 'border--focus--noTransition', 'transition-default', {
    'groupbtn--disabled': props.disabled,
    'groupbtn--active': isActive,
    'groupbtn--first': isFirstButton,
    'groupbtn--last': isLastButton,
  });

  return (
    <Wrapper>
      <Element {...rest} className={rootClass}>
        {Icon && (
          <div className="margin--right--half">
            <Icon size="xs" />
          </div>
        )}
        <span className="groupbtn-label">{label}</span>
      </Element>
    </Wrapper>
  );
};

GroupButton.defaultProps = {
  Wrapper: (props) => props.children,
};

GroupButton.propTypes = {
  /** Displays active button style when true */
  isActive: PropTypes.bool,

  /**
   * Takes in a react-router `Link` reference and sets it
   * as the base element. You may ONLY use it like the
   * following:
   *
   * - `import { Link } from 'react-router'`
   * - `Link={Link}`
   */
  Link: PropTypes.func,

  /** Text that appears in GroupButton */
  label: PropTypes.string,

  /** Controls whether button is disabled or not (works for
   * anchor tags as well, visually) */
  disabled: PropTypes.bool,

  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: PropTypes.func,

  /**
   * A component to wrap the entire group button. The
   * `wrapper` must render the children prop passed into it
   * to render the GroupButton.
   */
  Wrapper: PropTypes.elementType,

  /** Assigns the first button styling. */
  isFirstButton: PropTypes.bool,

  /** Assigns the last button styling. */
  isLastButton: PropTypes.bool,
};

export default GroupButton;
