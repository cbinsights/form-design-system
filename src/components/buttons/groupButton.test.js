import React from 'react';
import { shallow } from 'enzyme';

import GroupButton from './GroupButton';

const Icon = () => (
  <span>ðŸŒ­</span>
);
const Wrapper = (props) => (
  <div {...props} data-test="snapshot-wrapper" />
);

describe('GroupButton component', () => {
  it('matches snapshot (set all props)', () => {
    expect(shallow(<GroupButton value='Feed' Icon={Icon} isActive={true} Wrapper={Wrapper} />)).toMatchSnapshot();
  });
});

