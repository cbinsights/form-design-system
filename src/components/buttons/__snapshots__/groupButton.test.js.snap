// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupButton component matches snapshot (set all props) 1`] = `
<Wrapper>
  <button
    className="groupbtn border--focus--noTransition transition-default groupbtn--active"
    value="Feed"
  >
    <div
      className="margin--right--half"
    >
      <Icon
        size="xs"
      />
    </div>
    <span
      className="groupbtn-label"
    />
  </button>
</Wrapper>
`;
